go/src/oauth2client/oauth2client.go

```diff
@@ -0,0 +1,285 @@
+package oauth2client
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const (
+	// The URN for getting verification token offline
+	oobCallbackUrn = "urn:ietf:wg:oauth:2.0:oob"
+	// The URN for token request grant type jwt-bearer
+	jwtBearerUrn = "urn:ietf:params:oauth:grant-type:jwt-bearer"
+)
+
+// handle 3LO authorize flow. prints the authorization URL on stdout and reads
+// the verification code form stdin.
+func defaultAuthorizeFlowHandler(authorizeUrl string) string {
+	// Print the url on console, let user authorize and paste the token back.
+	fmt.Printf("Go to the following link in your browser:\n\n   %s\n\n", authorizeUrl)
+
+	fmt.Println("Enter verification code: ")
+	code := ""
+	fmt.Scanln(&code)
+	return code
+}
+
+// Run the three-legged oauth authorize flow.
+func authorizeFlow(secret map[string]interface{}, scopes string, handler func(string) string) string {
```

Comment:
Use "scope string" everywhere in this file, because that is OAuth 2 spec. The scopes []string should only be used in the main.go.